### PR TITLE
Fixed #27402 - LocaleMiddleware redirecting when it should display 404.

### PR DESCRIPTION
**Pull request by hovi**

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/middleware/locale.py [hunk 1]
```diff
@@ -37,6 +37,8 @@ def process_response(self, request, response):
 
         if response.status_code == 404 and not language_from_path and i18n_patterns_used:
             language_path = '/%s%s' % (language, request.path_info)
+            if i18n_patterns_used and not prefixed_default_language and str(language) == str(settings.LANGUAGE_CODE):
```
Comment (francoisfreitag) [inline django/middleware/locale.py]: There is no need to check `i18n_patterns_used` here, it's already checked above.
I believe the cast to `str()` is unnecessary.

File: tests/i18n/tests.py [hunk 2]
```diff
@@ -1859,6 +1859,10 @@ def test_page_with_dash(self):
         response = self.client.get('/de-simple-page/')
         self.assertEqual(response.content, b'Yes')
 
+    def test_404_page_no_default_language_redirect(self):
+        response = self.client.get('/non-existent-no-slash', follow=False)
```
Comment (francoisfreitag) [inline tests/i18n/tests.py]: I don't think `follow` is necessary here, since you should receive a 404.

File: django/middleware/locale.py [hunk 3]
```diff
@@ -37,6 +37,8 @@ def process_response(self, request, response):
...
+                language_path = '/%s' % (request.path_info)
```
Comment (francoisfreitag) [inline django/middleware/locale.py]: IMHO, it's clearer if you use:
```python
if not prefixed_default_language and language == settings.LANGUAGE_CODE:
    language_path = '/%s' % (request.path_info)
else:
    language_path = '/%s%s' % (language, request.path_info)
```

File: tests/i18n/tests.py (hunk 2 above)
Comment (francoisfreitag) [inline tests/i18n/tests.py]: On second thought, I'm actually in favour of expliciting follow=False, it prevents the test to pass if the redirection to the 404 page occurs.

Comment (hovi) [inline tests/i18n/tests.py]: Yeah that is exactly why I kept it.
Another option is to remove follow and add assert that response.redirect_chain is None, that may be more expressive to what is supposed to be tested.

File: django/middleware/locale.py (hunk 1 above)
Comment (hovi) [inline django/middleware/locale.py]: Yeah.

**Comment by francoisfreitag** [comment]
The middleware actually tries to redirect the user request to the internationalized url when a 404 is generated, for cases where the user sent a valid request without specifying the language code in the request path. ([here](https://github.com/django/django/blob/master/django/middleware/locale.py#L39-L57))
This behavior seems correct to me.

What you actually want to prevent is the redirection by the middleware [here](https://github.com/django/django/blob/master/django/middleware/locale.py#L49-L57) when you are using `prefix_default_language=False` and the request in the default language.

I believe a better approach would be something similar to replacing [the `if` before the redirect bloc](https://github.com/django/django/blob/master/django/middleware/locale.py#L48) with:
```python
if (path_valid or path_needs_slash) and (  # The url actually exists if the language is prefixed to the query
    not prefixed_default_language and language == settings.LANGUAGE_CODE):  # You can redirect to the prefixed version of the URL
```

**Comment by timgraham** [comment]
Continued in https://github.com/django/django/pull/7483.

**Pull request closed without merging**